xla/service/gpu/gpu_compiler_test.cc

```diff
@@ -1233,6 +1233,52 @@ ENTRY main {
+  %Arg_1.5 = f32[32,32]{1,0} parameter(1), metadata={op_name="jit(h)/jit(main)/pjit" scheduling_name="Arg_1.5"}
+  %Arg_0.4 = f32[32,32]{1,0} parameter(0), metadata={op_name="jit(h)/jit(main)/pjit" scheduling_name="Arg_0.4"}
+  %custom-call.1 = (f32[32,32]{1,0}, s8[8192]{0}) custom-call(f32[32,32]{1,0} %Arg_0.4, f32[32,32]{1,0} %Arg_1.5), custom_call_target="__cublas$gemm", metadata={op_name="jit(h)/jit(main)/jit(f)/dot_general" source_file="/mounted/scheduling-example/tiny.py" source_line=17 scheduling_name="custom-call.1"}, backend_config={"operation_queue_id":"0","wait_on_operation_queues":[],"gemm_backend_config":{"alpha_real":1,"alpha_imag":0,"beta":0,"dot_dimension_numbers":{"lhs_contracting_dimensions":["1"],"rhs_contracting_dimensions":["0"],"lhs_batch_dimensions":[],"rhs_batch_dimensions":[]},"precision_config":{"operand_precision":["DEFAULT","DEFAULT"],"algorithm":"ALG_UNSET"},"epilogue":"DEFAULT","damax_output":false,"lhs_stride":"1024","rhs_stride":"1024","grad_x":false,"grad_y":false},"force_earliest_schedule":false}
+  ROOT %get-tuple-element = f32[32,32]{1,0} get-tuple-element((f32[32,32]{1,0}, s8[8192]{0}) %custom-call.1), index=0, metadata={op_name="jit(h)/jit(main)/jit(f)/dot_general" source_file="/mounted/scheduling-example/tiny.py" source_line=17 scheduling_name="get-tuple-element"}
+}, execution_thread="explicit"
+
+ENTRY %main.8 (Arg_0.1.0: f32[32,32], Arg_1.2.0: f32[32,32]) -> f32[32,32] {
+  %Arg_1.2.0 = f32[32,32]{1,0} parameter(1), metadata={op_name="b" scheduling_name="Arg_1.2.0"}
+  %Arg_0.1.0 = f32[32,32]{1,0} parameter(0), metadata={op_name="a" scheduling_name="Arg_0.1.0"}
+  %call-start = ((f32[32,32]{1,0}, f32[32,32]{1,0}), f32[32,32]{1,0}) call-start(f32[32,32]{1,0} %Arg_0.1.0, f32[32,32]{1,0} %Arg_1.2.0), async_execution_thread="explicit", to_apply=%f.3, frontend_attributes={_xla_stream_annotation="1"}, metadata={scheduling_name="call-start"}
+  ROOT %call-done = f32[32,32]{1,0} call-done(((f32[32,32]{1,0}, f32[32,32]{1,0}), f32[32,32]{1,0}) %call-start), frontend_attributes={_xla_stream_annotation="1"}, metadata={scheduling_name="call-done"}, backend_config={"operation_queue_id":"0","wait_on_operation_queues":[],"force_earliest_schedule":false}
+})";
+  auto module = ParseAndReturnVerifiedModule(hlo_text).value();
+
+  std::unique_ptr<Executable> executable =
+      backend()
+          .compiler()
+          ->RunBackend(std::move(module), backend().default_stream_executor(),
+                       {/*device_allocator=*/nullptr,
+                        /*thread_pool=*/nullptr,
+                        /*layout_canonicalization_callback=*/{},
+                        /*is_autotuning_compilation=*/false})
+          .value();
+  std::unique_ptr<GpuExecutable> gpu_exec(
+      static_cast<GpuExecutable*>(executable.release()));
+  // We expect there to be only 3 thunks:
+  // 1) wait 2) a sequential thunk and 3) another wait.
+  EXPECT_EQ(gpu_exec->GetThunk().thunks().size(), 3);
```

Comment:
Nit: `EXPECT_EQ(gpu_exec->GetThunk().thunks(), testing::SizeIs(3))` is typically preferred.

Alternatively, I think this test could be improved with a matcher, something like this:

```
MATCHER_P(ThunkKindIs, kind, "") {
  return ExplainMatchResult(Eq(kind), arg->kind(), result_listener);
}

EXPECT_THAT(gpu_exec->GetThunk().thunks(), ElementsAre(ThunkTypeIs(...), ThunkTypeIs(...), ThunkTypeIs(...)));
```

IMO the latter is more readable and has the advantage of testing all the thunk types, not just the middle one.

@bchetioui opinions?

Reply:
Good catch. I'd also be happiest with the matcher!

Reply:
Added